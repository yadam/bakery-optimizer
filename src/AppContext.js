import PropTypes from 'prop-types';
import React, { useState } from 'react';

export const AppContext = React.createContext();

export const AppContextProvider = ({ children }) => {
  const [breads, setBreads] = useState([]);
  const [customers, setCustomers] = useState([]);
  return (
    <AppContext.Provider value={{ breads, customers, setBreads, setCustomers }}>
      {children}
    </AppContext.Provider>
  );
};

AppContextProvider.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.arrayOf(PropTypes.node),
  ]),
};

AppContextProvider.defaultProps = {
  children: undefined,
};

export default { AppContext, AppContextProvider };
