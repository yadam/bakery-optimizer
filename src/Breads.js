import React, { memo, useContext, useState } from 'react';
import {
  Box,
  Button,
  Chip,
  Grid,
  TextField,
  Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { AppContext } from './AppContext';

const useStyles = makeStyles((theme) => ({
  root: {
    margin: theme.spacing(6, 0, 3),
  },
  input: {
    margin: theme.spacing(2, 0, 0, 0),
  },
  list: {
    padding: theme.spacing(0, 1, 1, 1),
  },
  button: {
    margin: theme.spacing(1.25, 0, 0, 1),
  },
}));

export const Breads = memo(() => {
  const { breads, setBreads } = useContext(AppContext);
  const [breadInput, setBreadInput] = useState('');
  const classes = useStyles();

  const addBread = () => {
    if (breadInput && !breads.includes(breadInput)) {
      setBreads(breads.concat([breadInput]));
    }

    // clear the input field
    setBreadInput('');
  };

  const onChangeHandler = (e) => setBreadInput(e.target.value);

  const handleKeyPress = (e) => {
    if (e.key === 'Enter') {
      addBread();
    }
  };

  return (
    <Box className={classes.root}>
      <Typography className={classes.list}>Bread types:</Typography>
      {breads.map((bread) => (
        <Chip key={bread} label={bread} />
      ))}
      <Grid container className={classes.input}>
        <Grid xs={10} item>
          <TextField
            id="new-bread"
            label="Enter Bread Type"
            variant="outlined"
            value={breadInput}
            onChange={onChangeHandler}
            onKeyPress={handleKeyPress}
            fullWidth
          />
        </Grid>
        <Grid xs={2} item>
          <Button
            variant="contained"
            color="primary"
            className={classes.button}
            onClick={addBread}
            fullWidth
          >
            Add
          </Button>
        </Grid>
      </Grid>
    </Box>
  );
});

export default Breads;
