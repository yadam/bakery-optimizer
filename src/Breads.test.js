import React from 'react';
import { act, create } from 'react-test-renderer';
import { AppContext } from './AppContext';
import { Breads } from './Breads';

const breads = ['sourdough', 'wheat', 'banana'];

describe('Breads', () => {
  it('renders', () => {
    const setBreadsMock = jest.fn();
    const tree = create(
      <AppContext.Provider value={{ breads: [], setBreads: setBreadsMock }}>
        <Breads />
      </AppContext.Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with breads', () => {
    const setBreadsMock = jest.fn();
    const tree = create(
      <AppContext.Provider value={{ breads, setBreads: setBreadsMock }}>
        <Breads />
      </AppContext.Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('adds a bread', () => {
    const setBreadsMock = jest.fn();
    const tree = create(
      <AppContext.Provider value={{ breads, setBreads: setBreadsMock }}>
        <Breads />
      </AppContext.Provider>,
    );
    const instance = tree.root;
    const input = instance.findByType('input');
    const button = instance.findByType('button');
    act(() => {
      input.props.onChange({ target: { value: 'brioche' } });
    });
    act(() => {
      button.props.onClick();
    });
    expect(setBreadsMock).toHaveBeenCalledTimes(1);
  });
});
