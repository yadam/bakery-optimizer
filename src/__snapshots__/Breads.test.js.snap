// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breads renders 1`] = `
<div
  className="MuiBox-root MuiBox-root-5 makeStyles-root-1"
>
  <p
    className="MuiTypography-root makeStyles-list-3 MuiTypography-body1"
  >
    Bread types:
  </p>
  <div
    className="MuiGrid-root makeStyles-input-2 MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
          htmlFor="new-bread"
          id="new-bread-label"
        >
          Enter Bread Type
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="new-bread"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-8"
            >
              <span>
                Enter Bread Type
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-4 MuiButton-containedPrimary MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Add
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Breads renders with breads 1`] = `
<div
  className="MuiBox-root MuiBox-root-10 makeStyles-root-1"
>
  <p
    className="MuiTypography-root makeStyles-list-3 MuiTypography-body1"
  >
    Bread types:
  </p>
  <div
    className="MuiChip-root"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      className="MuiChip-label"
    >
      sourdough
    </span>
  </div>
  <div
    className="MuiChip-root"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      className="MuiChip-label"
    >
      wheat
    </span>
  </div>
  <div
    className="MuiChip-root"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
  >
    <span
      className="MuiChip-label"
    >
      banana
    </span>
  </div>
  <div
    className="MuiGrid-root makeStyles-input-2 MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-10"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        onKeyPress={[Function]}
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink={false}
          htmlFor="new-bread"
          id="new-bread-label"
        >
          Enter Bread Type
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="new-bread"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-6 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-8"
            >
              <span>
                Enter Bread Type
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-button-4 MuiButton-containedPrimary MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Add
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
